tuna/export_db.py

```diff
@@ -95,237 +105,240 @@ def parse_args():
-      kinder = kern_db.get(key)
-      if kinder:
-        continue
-      num_kdb_blobs += 1
-      kern_db[key] = (blob.kernel_blob, blob.kernel_hash,
-                      blob.uncompressed_size)
 
-  LOGGER.warning("Total number of FDB entries: %s", num_fdb_entries)
-  LOGGER.warning("Total number of blobs: %s", num_kdb_blobs)
-  return kern_db
+def get_pdb_query(dbt, args):
+  """Compose query to get perf_db rows based on filters from args"""
+  src_table = dbt.find_db_table
+  if args.golden_v is not None:
+    src_table = dbt.golden_table
+  query = get_fdb_query(dbt, args)
+  query = query.filter(dbt.solver_table.tunable == 1)\
+        .filter(src_table.params != '')
 
+  return query
 
-def export_kdb(dbt, args):
-  """
-  Function to export the kernel cache
-  """
-  query = fdb_query(dbt, args)
 
+def get_fdb_alg_lists(query):
+  """return dict with key: fdb_key + alg_lib, val: solver list"""
   find_db = OrderedDict()
   solvers = {}
-  for fdb_entry, _ in query.all():
+  for fdb_entry, _, _ in query.all():
     fdb_key = fdb_entry.fdb_key
     if fdb_key not in solvers:
       solvers[fdb_key] = {}
-    #skip if there is a solver for this key
     if fdb_entry.solver in solvers[fdb_key].keys():
-      #LOGGER.warning("skipped duplicate solver: %s : %s : %s vs stored: %s", fdb_key,
-      #               fdb_entry.solver, fdb_entry.update_ts,
-      #               solvers[fdb_key][fdb_entry.solver])
+      LOGGER.warning("Skipped duplicate solver: %s : %s with ts %s vs prev %s",
+                     fdb_key, fdb_entry.solver, fdb_entry.update_ts,
+                     solvers[fdb_key][fdb_entry.solver])
       continue
-    #record found solver for key
     solvers[fdb_key][fdb_entry.solver] = fdb_entry.update_ts
-    lst = find_db.get(fdb_key)
+
+    fdb_key_alg = (fdb_entry.fdb_key, fdb_entry.alg_lib)
+    lst = find_db.get(fdb_key_alg)
     if not lst:
-      find_db[fdb_key] = [fdb_entry]
+      find_db[fdb_key_alg] = [fdb_entry]
     else:
       lst.append(fdb_entry)
 
-  LOGGER.info("Building kdb.")
-  kern_db = build_miopen_kdb(find_db)
-
-  LOGGER.info("write kdb to file.")
-  return write_kdb(args.arch, args.num_cu, kern_db, args.filename)
+  return find_db
 
 
-def get_filename(arch, num_cu, filename, ocl, db_type):
-  """Helper function to compose filename"""
-  version = "1.0.0"
-  tuna_dir = f'tuna_{version}'
-  if not os.path.exists(tuna_dir):
-    os.makedirs(tuna_dir)
-  final_name = f"{tuna_dir}/{arch}_{num_cu}"
-  if num_cu > 64:
-    final_name = f'{tuna_dir}/{arch}{num_cu:x}'
-  if filename:
-    final_name = f'{tuna_dir}/{filename}'
+def build_miopen_fdb(fdb_alg_lists):
+  """ create miopen find db object for export
+  """
+  num_fdb_entries = 0
+  miopen_fdb = OrderedDict()
+  for fdbkey_alg, alg_entries in fdb_alg_lists.items():
+    fdb_key = fdbkey_alg[0]
+    num_fdb_entries += 1
+    #pick fastest solver for each algorithm
+    alg_entries.sort(key=lambda x: float(x.kernel_time))
```

Comment:
we can do this in mysql if we order by kernel_time desc in the query building function. I would recommend we do that there.

Reply:
Query that originally collected this entry was across all configs + solvers in a session (or golden_v, arch, num_cu).  

After the query the resulting list of db objects is sorted and mapped by a fdb_key + alg tuple. (There is also some logic removing duplicate solvers by update timestamp, which may no longer be relevant).

The sort that you see here is across solvers in one of these fdb_key + alg mappings. Of which we keep the fastest solver.

If we sort by runtime in sql, we will still need to hand pick out duplicate algorithms for a config + session. I don't see the benefit in restructuring this way.